File: src/memorygraph/integration/project_analysis.py
"""
Project-Aware Memory for Claude Code Integration.

Provides project detection, codebase analysis, and file change tracking:
- Project detection from directory structure
- Codebase analysis (languages, frameworks, structure)
- File change tracking with git integration
- Code pattern identification
"""

import json
import os
import re
import subprocess
from collections import Counter
from datetime import datetime
from pathlib import Path
from typing import Any, Optional
from uuid import uuid4

from pydantic import BaseModel, Field

from ..backends.base import GraphBackend


class ProjectInfo(BaseModel):
    """Project information."""

    project_id: str = Field(default_factory=lambda: str(uuid4()))
    name: str = Field(..., description="Project name")
    path: str = Field(..., description="Project directory path")
    project_type: str = Field(..., description="Project type (e.g., 'python', 'typescript', 'mixed')")
    git_remote: Optional[str] = Field(None, description="Git remote URL if available")
    description: Optional[str] = Field(None, description="Project description")
    technologies: list[str] = Field(default_factory=list, description="Technologies used")


class CodebaseInfo(BaseModel):
    """Codebase analysis results."""

    total_files: int = Field(..., description="Total number of files")
    file_types: dict[str, int] = Field(
        default_factory=dict, description="File count by extension"
    )
    languages: list[str] = Field(default_factory=list, description="Programming languages detected")
    frameworks: list[str] = Field(default_factory=list, description="Frameworks detected")
    structure: dict[str, Any] = Field(default_factory=dict, description="Directory structure")
    config_files: list[str] = Field(default_factory=list, description="Configuration files found")


class FileChange(BaseModel):
    """File change information."""

    file_path: str = Field(..., description="Path to changed file")
    change_type: str = Field(..., description="Type of change (added, modified, deleted)")
    timestamp: datetime = Field(default_factory=datetime.now)
    lines_added: int = Field(default=0, description="Lines added")
    lines_removed: int = Field(default=0, description="Lines removed")


class Pattern(BaseModel):
    """Code pattern identified."""

    pattern_id: str = Field(default_factory=lambda: str(uuid4()))
    pattern_type: str = Field(..., description="Type of pattern")
    description: str = Field(..., description="Pattern description")
    examples: list[str] = Field(default_factory=list, description="Example occurrences")
    frequency: int = Field(default=1, description="Frequency of pattern")
    confidence: float = Field(default=0.5, description="Confidence in pattern")


# File patterns to ignore
IGNORE_PATTERNS = [
    ".git",
    ".svn",
    ".hg",
    "node_modules",
    "__pycache__",
    ".pytest_cache",
    ".mypy_cache",
    ".tox",
    "venv",
    ".venv",
    "env",
    ".env",
    "dist",
    "build",
    "*.pyc",
    "*.pyo",
    "*.egg-info",
    ".DS_Store",
    "thumbs.db",
]

# Config file patterns for project type detection
PROJECT_CONFIGS = {
    "python": ["pyproject.toml", "setup.py", "requirements.txt", "Pipfile", "poetry.lock"],
    "typescript": ["package.json", "tsconfig.json", "yarn.lock", "pnpm-lock.yaml"],
    "javascript": ["package.json", "yarn.lock", "pnpm-lock.yaml"],
    "rust": ["Cargo.toml", "Cargo.lock"],
    "go": ["go.mod", "go.sum"],
    "java": ["pom.xml", "build.gradle", "build.gradle.kts"],
    "ruby": ["Gemfile", "Gemfile.lock"],
    "php": ["composer.json", "composer.lock"],
}

# Framework detection patterns
FRAMEWORK_PATTERNS = {
    "react": ["react", "@types/react"],
    "vue": ["vue", "@vue/"],
    "angular": ["@angular/"],
    "next": ["next", "next.config"],
    "fastapi": ["fastapi"],
    "flask": ["flask"],
    "django": ["django"],
    "express": ["express"],
    "nestjs": ["@nestjs/"],
    "spring": ["spring-boot", "springframework"],
}


async def detect_project(backend: GraphBackend, directory: str) -> Optional[ProjectInfo]:
    """
    Detect project from directory and return project information.

    Args:
        backend: Database backend
        directory: Directory path to analyze

    Returns:
        ProjectInfo if project detected, None otherwise

    Example:
        >>> project = await detect_project(backend, "/Users/me/my-app")
        >>> print(project.name, project.project_type)
    """
    directory = os.path.abspath(os.path.expanduser(directory))

    if not os.path.isdir(directory):
        return None

    # Extract project name from directory
    project_name = os.path.basename(directory)

    # Check for git remote
    git_remote = None
    try:
        result = subprocess.run(
            ["git", "-C", directory, "config", "--get", "remote.origin.url"],
            capture_output=True,
            text=True,
            timeout=5,
        )
        if result.returncode == 0:
            git_remote = result.stdout.strip()
            # Extract repo name from git URL if available
            if git_remote:
                match = re.search(r"[/:]([^/]+?)(?:\.git)?$", git_remote)
                if match:
                    project_name = match.group(1)
    except Exception:
        pass

    # Detect project type from config files
    project_type = "unknown"
    config_files = []
    technologies = []

    for lang, configs in PROJECT_CONFIGS.items():
        for config in configs:
            config_path = os.path.join(directory, config)
            if os.path.isfile(config_path):
                config_files.append(config)
                if project_type == "unknown":
                    project_type = lang

                # Parse config file for more details
                if config == "package.json":
                    try:
                        with open(config_path) as f:
                            package_data = json.load(f)
                            dependencies = {
                                **package_data.get("dependencies", {}),
                                **package_data.get("devDependencies", {}),
                            }

                            # Detect frameworks
                            for framework, patterns in FRAMEWORK_PATTERNS.items():
                                if any(p in dep for dep in dependencies for p in patterns):
                                    technologies.append(framework)
                    except Exception:
                        pass

                elif config == "pyproject.toml":
                    try:
                        with open(config_path) as f:
                            content = f.read()
                            # Detect Python frameworks
                            for framework, patterns in FRAMEWORK_PATTERNS.items():
                                if any(p in content for p in patterns):
                                    technologies.append(framework)
                    except Exception:
                        pass

    # If multiple config types found, mark as mixed (but handle TypeScript special case)
    detected_types = [
        lang for lang, configs in PROJECT_CONFIGS.items() if any(c in config_files for c in configs)
    ]

    # Special case: if tsconfig.json is present, it's TypeScript (not mixed with JavaScript)
    # because TypeScript projects also use package.json
    if "tsconfig.json" in config_files:
        project_type = "typescript"
        detected_types = ["typescript"]
    elif len(detected_types) > 1:
        project_type = "mixed"
        technologies.extend(detected_types)

    project = ProjectInfo(
        name=project_name,
        path=directory,
        project_type=project_type,
        git_remote=git_remote,
        technologies=list(set(technologies)),  # Remove duplicates
    )

    # Check if project already exists in database
    existing = await backend.search_nodes(
        "Entity", {"type": "project", "name": project_name, "path": directory}
    )

    if existing:
        project.project_id = existing[0]["id"]
        # Update existing project
        await backend.execute_query(
            """
            MATCH (p:Entity {id: $project_id})
            SET p.git_remote = $git_remote,
                p.project_type = $project_type,
                p.technologies = $technologies,
                p.updated_at = datetime()
            """,
            {
                "project_id": project.project_id,
                "git_remote": git_remote,
                "project_type": project_type,
                "technologies": technologies,
            },
        )
    else:
        # Create new project entity
        properties = {
            "id": project.project_id,
            "type": "project",
            "name": project_name,
            "path": directory,
            "project_type": project_type,
            "git_remote": git_remote,
            "technologies": technologies,
            "created_at": datetime.now(),
            "updated_at": datetime.now(),
        }
        await backend.store_node("Entity", properties)

    return project


async def analyze_codebase(backend: GraphBackend, directory: str) -> CodebaseInfo:
    """
    Analyze codebase structure and characteristics.

    Args:
        backend: Database backend
        directory: Directory path to analyze

    Returns:
        CodebaseInfo with analysis results

    Example:
        >>> info = await analyze_codebase(backend, "/Users/me/my-app")
        >>> print(f"Total files: {info.total_files}")
        >>> print(f"Languages: {', '.join(info.languages)}")
    """
    directory = os.path.abspath(os.path.expanduser(directory))

    file_types: Counter = Counter()
    config_files = []
    total_files = 0

    # Walk directory tree
    for root, dirs, files in os.walk(directory):
        # Filter out ignored directories
        dirs[:] = [d for d in dirs if d not in IGNORE_PATTERNS and not d.startswith(".")]

        for file in files:
            # Skip ignored patterns
            if any(file.endswith(pattern.replace("*", "")) for pattern in IGNORE_PATTERNS):
                continue

            total_files += 1
            ext = os.path.splitext(file)[1]
            if ext:
                file_types[ext] += 1

            # Check if it's a config file
            for configs in PROJECT_CONFIGS.values():
                if file in configs:
                    config_files.append(os.path.join(root, file))

    # Map extensions to languages
    ext_to_lang = {
        ".py": "python",
        ".js": "javascript",
        ".ts": "typescript",
        ".tsx": "typescript",
        ".jsx": "javascript",
        ".rs": "rust",
        ".go": "go",
        ".java": "java",
        ".rb": "ruby",
        ".php": "php",
        ".c": "c",
        ".cpp": "cpp",
        ".h": "c",
        ".hpp": "cpp",
        ".cs": "csharp",
        ".swift": "swift",
        ".kt": "kotlin",
    }

    languages = list({ext_to_lang.get(ext, "other") for ext in file_types if ext in ext_to_lang})

    # Detect frameworks from file analysis
    frameworks = []

    codebase_info = CodebaseInfo(
        total_files=total_files,
        file_types=dict(file_types),
        languages=languages,
        frameworks=frameworks,
        config_files=config_files,
    )

    return codebase_info


async def track_file_changes(
    backend: GraphBackend, repo_path: str, project_id: str
) -> list[FileChange]:
    """
    Track file changes using git diff.

    Args:
        backend: Database backend
        repo_path: Path to git repository
        project_id: Project ID

    Returns:
        List of FileChange objects

    Example:
        >>> changes = await track_file_changes(
        ...     backend,
        ...     "/Users/me/my-app",
        ...     "project-123"
        ... )
        >>> for change in changes:
        ...     print(f"{change.change_type}: {change.file_path}")
    """
    repo_path = os.path.abspath(os.path.expanduser(repo_path))

    changes = []

    try:
        # Get git status for changed files
        result = subprocess.run(
            ["git", "-C", repo_path, "status", "--porcelain"],
            capture_output=True,
            text=True,
            timeout=10,
        )

        if result.returncode != 0:
            return changes

        for line in result.stdout.strip().split("\n"):
            if not line:
                continue

            status = line[:2].strip()
            file_path = line[3:].strip()

            # Map git status to change type
            change_type = "modified"
            if status in ["A", "??"]:
                change_type = "added"
            elif status == "D":
                change_type = "deleted"
            elif status in ["M", "MM"]:
                change_type = "modified"

            # Get diff stats for modified files
            lines_added = 0
            lines_removed = 0

            if change_type == "modified" and os.path.isfile(os.path.join(repo_path, file_path)):
                try:
                    diff_result = subprocess.run(
                        ["git", "-C", repo_path, "diff", "--numstat", "HEAD", file_path],
                        capture_output=True,
                        text=True,
                        timeout=5,
                    )
                    if diff_result.returncode == 0 and diff_result.stdout:
                        parts = diff_result.stdout.strip().split("\t")
                        if len(parts) >= 2:
                            lines_added = int(parts[0]) if parts[0].isdigit() else 0
                            lines_removed = int(parts[1]) if parts[1].isdigit() else 0
                except Exception:
                    pass

            file_change = FileChange(
                file_path=file_path,
                change_type=change_type,
                lines_added=lines_added,
                lines_removed=lines_removed,
            )
            changes.append(file_change)

            # Store file change as observation
            properties = {
                "id": str(uuid4()),
                "type": "file_change",
                "title": f"File {change_type}: {file_path}",
                "content": f"File: {file_path}\nChange: {change_type}\n"
                f"Lines added: {lines_added}\nLines removed: {lines_removed}",
                "context": {
                    "file_path": file_path,
                    "change_type": change_type,
                    "lines_added": lines_added,
                    "lines_removed": lines_removed,
                    "project_id": project_id,
                },
                "created_at": datetime.now(),
                "updated_at": datetime.now(),
            }

            memory_id = await backend.store_node("Memory", properties)

            # Link to project
            await backend.store_relationship(
                memory_id,
                project_id,
                "PART_OF",
                {"created_at": datetime.now(), "strength": 1.0},
            )

            # Create or get file entity and link
            file_entity = await backend.execute_query(
                """
                MERGE (f:Entity {name: $file_path, type: 'file'})
                ON CREATE SET f.id = $file_id, f.created_at = datetime()
                RETURN f.id as id
                """,
                {"file_path": file_path, "file_id": str(uuid4())},
            )

            if file_entity:
                file_id = file_entity[0]["id"]
                await backend.store_relationship(
                    memory_id,
                    file_id,
                    "MODIFIES" if change_type == "modified" else "CREATES",
                    {"created_at": datetime.now(), "strength": 1.0},
                )

    except Exception as e:
        # Git not available or not a git repo - return empty list
        pass

    return changes


async def identify_code_patterns(
    backend: GraphBackend, project_id: str, files: list[str]
) -> list[Pattern]:
    """
    Identify code patterns in files.

    Args:
        backend: Database backend
        project_id: Project ID
        files: List of file paths to analyze

    Returns:
        List of identified patterns

    Example:
        >>> patterns = await identify_code_patterns(
        ...     backend,
        ...     "project-123",
        ...     ["src/api/users.py", "src/api/posts.py"]
        ... )
    """
    patterns: list[Pattern] = []

    # Common code patterns to identify
    pattern_regexes = {
        "api_endpoint": r"@(app|router)\.(get|post|put|delete|patch)\(['\"]([^'\"]+)",
        "class_definition": r"class\s+(\w+)(?:\(.*?\))?:",
        "function_definition": r"(?:async\s+)?def\s+(\w+)\s*\(",
        "import_statement": r"(?:from\s+[\w.]+\s+)?import\s+([\w, ]+)",
        "error_handling": r"try:|except\s+(\w+(?:Error|Exception)):",
        "async_await": r"\basync\s+def\b|\bawait\b",
        "type_annotation": r":\s*([A-Z][\w\[\], ]+)(?:\s*=)?",
    }

    pattern_counts: Counter = Counter()
    pattern_examples: dict[str, list[str]] = {}

    for file_path in files:
        if not os.path.isfile(file_path):
            continue

        try:
            with open(file_path, encoding="utf-8", errors="ignore") as f:
                content = f.read()

                for pattern_type, regex in pattern_regexes.items():
                    matches = re.findall(regex, content)
                    if matches:
                        pattern_counts[pattern_type] += len(matches)
                        if pattern_type not in pattern_examples:
                            pattern_examples[pattern_type] = []
                        # Store first few examples
                        pattern_examples[pattern_type].extend(
                            [str(m)[:100] for m in matches[:3]]
                        )
        except Exception:
            continue

    # Create pattern objects for significant patterns
    for pattern_type, count in pattern_counts.items():
        if count >= 2:  # Only patterns that occur at least twice
            pattern = Pattern(
                pattern_type=pattern_type,
                description=f"Code pattern: {pattern_type}",
                examples=pattern_examples.get(pattern_type, [])[:5],
                frequency=count,
                confidence=min(0.5 + (count * 0.05), 0.95),
            )

            # Store pattern as memory
            properties = {
                "id": pattern.pattern_id,
                "type": "code_pattern",
                "title": f"Pattern: {pattern_type}",
                "content": f"Pattern Type: {pattern_type}\n"
                f"Frequency: {count}\n\n"
                f"Examples:\n" + "\n".join(f"- {ex}" for ex in pattern.examples),
                "context": {
                    "pattern_type": pattern_type,
                    "frequency": count,
                    "confidence": pattern.confidence,
                    "project_id": project_id,
                },
                "created_at": datetime.now(),
                "updated_at": datetime.now(),
            }

            memory_id = await backend.store_node("Memory", properties)

            # Link to project
            await backend.store_relationship(
                memory_id,
                project_id,
                "FOUND_IN",
                {"created_at": datetime.now(), "strength": pattern.confidence},
            )

            patterns.append(pattern)

    return patterns

File: ts/src/integration/project-analysis.ts
/**
 * Project-Aware Memory for Claude Code Integration.
 *
 * Port of the Python `memorygraph.integration.project_analysis` module.
 *
 * Provides project detection, codebase analysis, and file change tracking:
 * - Project detection from directory structure
 * - Codebase analysis (languages, frameworks, structure)
 * - File change tracking with git integration
 * - Code pattern identification
 */

import { execSync } from "node:child_process";
import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { basename, extname, join, resolve } from "node:path";
import { randomUUID } from "node:crypto";

import type { GraphBackend } from "../backends/index.js";
import {
  createMemory,
  createRelationshipProperties,
  MemoryType,
  SearchQuerySchema,
  type Memory,
} from "../models.js";

// ---------------------------------------------------------------------------
// Interfaces
// ---------------------------------------------------------------------------

export interface ProjectInfo {
  project_id: string;
  name: string;
  path: string;
  project_type: string;
  git_remote?: string | null;
  description?: string | null;
  technologies: string[];
}

export interface CodebaseInfo {
  total_files: number;
  file_types: Record<string, number>;
  languages: string[];
  frameworks: string[];
  structure: Record<string, unknown>;
  config_files: string[];
}

export interface FileChange {
  file_path: string;
  change_type: string;
  timestamp: Date;
  lines_added: number;
  lines_removed: number;
}

export interface Pattern {
  pattern_id: string;
  pattern_type: string;
  description: string;
  examples: string[];
  frequency: number;
  confidence: number;
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const IGNORE_PATTERNS: string[] = [
  ".git",
  ".svn",
  ".hg",
  "node_modules",
  "__pycache__",
  ".pytest_cache",
  ".mypy_cache",
  ".tox",
  "venv",
  ".venv",
  "env",
  ".env",
  "dist",
  "build",
  ".DS_Store",
  "thumbs.db",
];

const PROJECT_CONFIGS: Record<string, string[]> = {
  python: ["pyproject.toml", "setup.py", "requirements.txt", "Pipfile", "poetry.lock"],
  typescript: ["package.json", "tsconfig.json", "yarn.lock", "pnpm-lock.yaml"],
  javascript: ["package.json", "yarn.lock", "pnpm-lock.yaml"],
  rust: ["Cargo.toml", "Cargo.lock"],
  go: ["go.mod", "go.sum"],
  java: ["pom.xml", "build.gradle", "build.gradle.kts"],
  ruby: ["Gemfile", "Gemfile.lock"],
  php: ["composer.json", "composer.lock"],
};

const FRAMEWORK_PATTERNS: Record<string, string[]> = {
  react: ["react", "@types/react"],
  vue: ["vue", "@vue/"],
  angular: ["@angular/"],
  next: ["next", "next.config"],
  fastapi: ["fastapi"],
  flask: ["flask"],
  django: ["django"],
  express: ["express"],
  nestjs: ["@nestjs/"],
  spring: ["spring-boot", "springframework"],
};

const EXT_TO_LANG: Record<string, string> = {
  ".py": "python",
  ".js": "javascript",
  ".ts": "typescript",
  ".tsx": "typescript",
  ".jsx": "javascript",
  ".rs": "rust",
  ".go": "go",
  ".java": "java",
  ".rb": "ruby",
  ".php": "php",
  ".c": "c",
  ".cpp": "cpp",
  ".h": "c",
  ".hpp": "cpp",
  ".cs": "csharp",
  ".swift": "swift",
  ".kt": "kotlin",
};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function buildMemory(
  type: string,
  title: string,
  content: string,
  metadata: Record<string, unknown>,
  id?: string
): Memory {
  return createMemory({
    id: id ?? randomUUID(),
    type,
    title,
    content,
    context: {
      additional_metadata: metadata,
    },
  });
}

function safeExecSync(cmd: string, opts: { cwd: string; timeout: number }): string | null {
  try {
    return execSync(cmd, { stdio: "pipe", ...opts }).toString().trim();
  } catch {
    return null;
  }
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Detect project from directory and return project information.
 */
export async function detectProject(
  backend: GraphBackend,
  directory: string
): Promise<ProjectInfo | null> {
  const dir = resolve(directory.replace(/^~/, process.env["HOME"] ?? "~"));

  let stat;
  try {
    stat = statSync(dir);
  } catch {
    return null;
  }
  if (!stat.isDirectory()) return null;

  // Extract project name from directory
  let projectName = basename(dir);

  // Check for git remote
  let gitRemote: string | null = null;
  const remoteResult = safeExecSync(
    `git -C "${dir}" config --get remote.origin.url`,
    { cwd: dir, timeout: 5000 }
  );
  if (remoteResult) {
    gitRemote = remoteResult;
    const match = /[/:]([^/]+?)(?:\.git)?$/.exec(gitRemote);
    if (match) projectName = match[1]!;
  }

  // Detect project type from config files
  let projectType = "unknown";
  const configFiles: string[] = [];
  const technologies: string[] = [];

  for (const [lang, configs] of Object.entries(PROJECT_CONFIGS)) {
    for (const config of configs) {
      const configPath = join(dir, config);
      if (existsSync(configPath) && statSync(configPath).isFile()) {
        configFiles.push(config);
        if (projectType === "unknown") projectType = lang;

        // Parse config file for more details
        if (config === "package.json") {
          try {
            const pkgData = JSON.parse(readFileSync(configPath, "utf-8"));
            const dependencies: Record<string, unknown> = {
              ...(pkgData["dependencies"] ?? {}),
              ...(pkgData["devDependencies"] ?? {}),
            };
            for (const [framework, patterns] of Object.entries(FRAMEWORK_PATTERNS)) {
              if (patterns.some((p) => Object.keys(dependencies).some((dep) => dep.includes(p)))) {
                technologies.push(framework);
              }
            }
          } catch {
            // ignore parse errors
          }
        } else if (config === "pyproject.toml") {
          try {
            const content = readFileSync(configPath, "utf-8");
            for (const [framework, patterns] of Object.entries(FRAMEWORK_PATTERNS)) {
              if (patterns.some((p) => content.includes(p))) {
                technologies.push(framework);
              }
            }
          } catch {
            // ignore read errors
          }
        }
      }
    }
  }

  // Determine detected types
  let detectedTypes = Object.entries(PROJECT_CONFIGS)
    .filter(([, configs]) => configs.some((c) => configFiles.includes(c)))
    .map(([lang]) => lang);

  // Special case: tsconfig.json means TypeScript
  if (configFiles.includes("tsconfig.json")) {
    projectType = "typescript";
    detectedTypes = ["typescript"];
  } else if (detectedTypes.length > 1) {
    projectType = "mixed";
    technologies.push(...detectedTypes);
  }

  const uniqueTechnologies = [...new Set(technologies)];

  const project: ProjectInfo = {
    project_id: randomUUID(),
    name: projectName,
    path: dir,
    project_type: projectType,
    git_remote: gitRemote,
    technologies: uniqueTechnologies,
  };

  // Check if project already exists in database
  let existing: Memory | null = null;
  try {
    const results = await backend.searchMemories(
      SearchQuerySchema.parse({
        memory_types: [MemoryType.PROJECT],
        limit: 100,
        include_relationships: false,
      })
    );
    existing =
      results.find(
        (m) =>
          m.title === projectName &&
          m.context?.additional_metadata?.["path"] === dir
      ) ?? null;
  } catch (err) {
    console.warn(`Failed to search for existing project:`, err);
  }

  if (existing) {
    project.project_id = existing.id!;
    try {
      await backend.executeQuery(
        `
        MATCH (p:Memory {id: $project_id})
        SET p.updated_at = datetime()
        RETURN p.id as id
        `,
        {
          project_id: project.project_id,
          git_remote: gitRemote,
          project_type: projectType,
          technologies: uniqueTechnologies,
        },
        true
      );
    } catch (err) {
      console.warn(`Failed to update existing project:`, err);
    }
  } else {
    const metadata: Record<string, unknown> = {
      name: projectName,
      path: dir,
      project_type: projectType,
      git_remote: gitRemote,
      technologies: uniqueTechnologies,
    };
    const memory = buildMemory(
      MemoryType.PROJECT,
      `Project: ${projectName}`,
      `Project: ${projectName}\nPath: ${dir}\nType: ${projectType}\nTechnologies: ${uniqueTechnologies.join(", ")}`,
      metadata,
      project.project_id
    );
    try {
      await backend.storeMemory(memory);
    } catch (err) {
      console.warn(`Failed to store new project:`, err);
    }
  }

  return project;
}

/**
 * Analyze codebase structure and characteristics.
 */
export async function analyzeCodebase(
  _backend: GraphBackend,
  directory: string
): Promise<CodebaseInfo> {
  const dir = resolve(directory.replace(/^~/, process.env["HOME"] ?? "~"));

  const fileTypes: Record<string, number> = {};
  const configFiles: string[] = [];
  let totalFiles = 0;

  // Walk directory tree
  walkDirectory(dir, (root, files) => {
    for (const file of files) {
      // Skip ignored patterns
      if (IGNORE_PATTERNS.some((p) => file === p || file.endsWith(p.replace(/\*/g, "")))) {
        continue;
      }
      totalFiles++;
      const ext = extname(file);
      if (ext) {
        fileTypes[ext] = (fileTypes[ext] ?? 0) + 1;
      }
      // Check if it's a config file
      for (const configs of Object.values(PROJECT_CONFIGS)) {
        if (configs.includes(file)) {
          configFiles.push(join(root, file));
        }
      }
    }
  });

  // Map extensions to languages
  const languages = [
    ...new Set(
      Object.keys(fileTypes)
        .filter((ext) => ext in EXT_TO_LANG)
        .map((ext) => EXT_TO_LANG[ext]!)
    ),
  ];

  return {
    total_files: totalFiles,
    file_types: fileTypes,
    languages,
    frameworks: [],
    structure: {},
    config_files: configFiles,
  };
}

/**
 * Recursively walk a directory, skipping ignored directories.
 */
function walkDirectory(
  root: string,
  callback: (root: string, files: string[]) => void
): void {
  let entries: import("node:fs").Dirent[];
  try {
    entries = readdirSync(root, { withFileTypes: true });
  } catch {
    return;
  }

  const files: string[] = [];
  const subdirs: string[] = [];

  for (const entry of entries) {
    const name = entry.name;
    // Filter out ignored directories
    if (entry.isDirectory()) {
      if (IGNORE_PATTERNS.includes(name) || name.startsWith(".")) continue;
      subdirs.push(join(root, name));
    } else if (entry.isFile()) {
      files.push(name);
    }
  }

  callback(root, files);
  for (const sub of subdirs) {
    walkDirectory(sub, callback);
  }
}

/**
 * Track file changes using git diff.
 */
export async function trackFileChanges(
  backend: GraphBackend,
  repoPath: string,
  projectId: string
): Promise<FileChange[]> {
  const dir = resolve(repoPath.replace(/^~/, process.env["HOME"] ?? "~"));
  const changes: FileChange[] = [];

  const statusOutput = safeExecSync(`git -C "${dir}" status --porcelain`, {
    cwd: dir,
    timeout: 10000,
  });
  if (!statusOutput) return changes;

  for (const line of statusOutput.split("\n")) {
    if (!line) continue;

    const status = line.slice(0, 2).trim();
    const filePath = line.slice(3).trim();

    // Map git status to change type
    let changeType = "modified";
    if (status === "A" || status === "??") {
      changeType = "added";
    } else if (status === "D") {
      changeType = "deleted";
    } else if (status === "M" || status === "MM") {
      changeType = "modified";
    }

    // Get diff stats for modified files
    let linesAdded = 0;
    let linesRemoved = 0;

    if (changeType === "modified" && existsSync(join(dir, filePath))) {
      const diffOutput = safeExecSync(
        `git -C "${dir}" diff --numstat HEAD -- "${filePath}"`,
        { cwd: dir, timeout: 5000 }
      );
      if (diffOutput) {
        const parts = diffOutput.split("\t");
        if (parts.length >= 2) {
          linesAdded = /^\d+$/.test(parts[0]!) ? parseInt(parts[0]!, 10) : 0;
          linesRemoved = /^\d+$/.test(parts[1]!) ? parseInt(parts[1]!, 10) : 0;
        }
      }
    }

    const fileChange: FileChange = {
      file_path: filePath,
      change_type: changeType,
      timestamp: new Date(),
      lines_added: linesAdded,
      lines_removed: linesRemoved,
    };
    changes.push(fileChange);

    // Store file change as observation
    const metadata: Record<string, unknown> = {
      file_path: filePath,
      change_type: changeType,
      lines_added: linesAdded,
      lines_removed: linesRemoved,
      project_id: projectId,
    };
    const memory = buildMemory(
      MemoryType.FILE_CONTEXT,
      `File ${changeType}: ${filePath}`,
      `File: ${filePath}\nChange: ${changeType}\nLines added: ${linesAdded}\nLines removed: ${linesRemoved}`,
      metadata
    );

    let memoryId: string;
    try {
      memoryId = await backend.storeMemory(memory);
    } catch (err) {
      console.warn(`Failed to store file change memory for ${filePath}:`, err);
      continue;
    }

    // Link to project
    try {
      await backend.createRelationship(
        memoryId,
        projectId,
        "PART_OF",
        createRelationshipProperties({ strength: 1.0 })
      );
    } catch (err) {
      console.warn(`Failed to link file change to project:`, err);
    }

    // Create or get file entity and link
    const fileId = randomUUID();
    try {
      await backend.executeQuery(
        `
        MERGE (f:Entity {name: $file_path, type: 'file'})
        ON CREATE SET f.id = $file_id, f.created_at = datetime()
        RETURN f.id as id
        `,
        { file_path: filePath, file_id: fileId },
        true
      );
      await backend.createRelationship(
        memoryId,
        fileId,
        changeType === "modified" ? "MODIFIES" : "CREATES",
        createRelationshipProperties({ strength: 1.0 })
      );
    } catch (err) {
      console.warn(`Failed to link file change to file entity:`, err);
    }
  }

  return changes;
}

/**
 * Identify code patterns in files.
 */
export async function identifyCodePatterns(
  backend: GraphBackend,
  projectId: string,
  files: string[]
): Promise<Pattern[]> {
  const patterns: Pattern[] = [];

  // Common code patterns to identify
  const patternRegexes: Record<string, RegExp> = {
    api_endpoint: /@(?:app|router)\.(?:get|post|put|delete|patch)\(['"]([^'"]+)/g,
    class_definition: /class\s+(\w+)(?:\(.*?\))?:/g,
    function_definition: /(?:async\s+)?def\s+(\w+)\s*\(/g,
    import_statement: /(?:from\s+[\w.]+\s+)?import\s+([\w, ]+)/g,
    error_handling: /try:|except\s+(\w+(?:Error|Exception)):/g,
    async_await: /\basync\s+def\b|\bawait\b/g,
    type_annotation: /:\s*([A-Z][\w\[\], ]+)(?:\s*=)?/g,
  };

  const patternCounts: Record<string, number> = {};
  const patternExamples: Record<string, string[]> = {};

  for (const filePath of files) {
    if (!existsSync(filePath)) continue;

    let content: string;
    try {
      content = readFileSync(filePath, "utf-8");
    } catch {
      continue;
    }

    for (const [patternType, regex] of Object.entries(patternRegexes)) {
      // Reset lastIndex for global regexes
      regex.lastIndex = 0;
      const matches = [...content.matchAll(regex)];
      if (matches.length > 0) {
        patternCounts[patternType] = (patternCounts[patternType] ?? 0) + matches.length;
        if (!(patternType in patternExamples)) patternExamples[patternType] = [];
        // Store first few examples
        for (const m of matches.slice(0, 3)) {
          patternExamples[patternType]!.push((m[0] ?? "").slice(0, 100));
        }
      }
    }
  }

  // Create pattern objects for significant patterns
  for (const [patternType, count] of Object.entries(patternCounts)) {
    if (count < 2) continue; // Only patterns that occur at least twice

    const confidence = Math.min(0.5 + count * 0.05, 0.95);
    const examples = (patternExamples[patternType] ?? []).slice(0, 5);

    const pattern: Pattern = {
      pattern_id: randomUUID(),
      pattern_type: patternType,
      description: `Code pattern: ${patternType}`,
      examples,
      frequency: count,
      confidence,
    };

    // Store pattern as memory
    const metadata: Record<string, unknown> = {
      pattern_type: patternType,
      frequency: count,
      confidence,
      project_id: projectId,
    };
    const memory = buildMemory(
      MemoryType.CODE_PATTERN,
      `Pattern: ${patternType}`,
      `Pattern Type: ${patternType}\nFrequency: ${count}\n\nExamples:\n` +
        examples.map((ex) => `- ${ex}`).join("\n"),
      metadata,
      pattern.pattern_id
    );

    let memoryId: string;
    try {
      memoryId = await backend.storeMemory(memory);
    } catch (err) {
      console.warn(`Failed to store code pattern ${patternType}:`, err);
      continue;
    }

    // Link to project
    try {
      await backend.createRelationship(
        memoryId,
        projectId,
        "FOUND_IN",
        createRelationshipProperties({ strength: confidence })
      );
    } catch (err) {
      console.warn(`Failed to link code pattern to project:`, err);
    }

    patterns.push(pattern);
  }

  return patterns;
}
